### PR TITLE
orchestra/run: py2 fix for copy_and_close

### PR DESCRIPTION
**Pull request by sebastian-philipp**
needed by  https://github.com/ceph/ceph/pull/35260
Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

### CODE REVIEW

**Comment by kshtsk** [comment]
Please no! Don't use io.StringIO for py2

**Comment by sebastian-philipp** [comment]
> Please no! Don't use io.StringIO for py2, use io.BytesIO on caller side.

fair enough. Will do. If I may ask: why?

**Comment by kshtsk** [comment]
> > Please no! Don't use io.StringIO for py2, use io.BytesIO on caller side.
> 
> fair enough. Will do. If I may ask: why?

Same question, why do you need to introduce unicode usage, where it is not required?
Also, it is very risky to make changes like this to core functionality which is tested only well for bytes on py2, I can understand why you guys might want to use StringIO on py3 since it is standard type `str`, but for py2 it is unicode, which makes only troubles.

**Comment by sebastian-philipp** [comment]
Cause the table of supported inputs of `copy_and_close` looks like so:

| *        | str           | unicode      | bytes       |
| -------- | ------------- | ------------ | ----------- |
| Py3      | supported     |  same as str | supported   |
| Py2      | supported     |  **missing** | supported   |

This PR just fills the missing gap

**Comment by kshtsk** [comment]
> Cause the table of supported inputs of `copy_and_close` looks like so:
> * 	str 	unicode 	bytes
> Py3 	supported 	same as str 	supported
> Py2 	supported 	**missing** 	supported
> 
> This PR just fills the missing gap

We do not need to support py2 code, and we don't want to spend much time for this, it is now only in maintenance mode, it highly desired to switch all branches to py3.
The unicode py2 is something ugly thing, you can't even check it's type correctly:
```
>>> isinstance('a', str)
True
>>> isinstance('a', bytes)
False
>>> isinstance(u'a', bytes)
False
>>> isinstance(u'a', str)
True
>>> isinstance(u'a', unicode)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'unicode' is not defined

```

**Comment by kshtsk** [comment]
> This PR just fills the missing gap

it is only introduce more troubles than you think, for example, you need to fix the tests, this will increase divergence with py3 branch, which will complicate backports, I'm not sure you will want to spend time to support this branch?

**Comment by kshtsk** [comment]
@sebastian-philipp if you want to go this way and take responsibility for supporting py2 branch. Here you go. Removing DNM label ;-)

File: teuthology/orchestra/test/test_run.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 from mock import MagicMock, patch
 from pytest import raises
 
-from six import ensure_str, ensure_binary
+from six import ensure_str, ensure_binary, PY2
```
Comment (kshtsk) [inline teuthology/orchestra/test/test_run.py:9]: The PY2 should go from teuthology.util.compat as it is in the run.

**Comment by sebastian-philipp** [comment]
> @sebastian-philipp if you want to go this way and take responsibility for supporting py2 branch. Here you go. Removing DNM label ;-)

Ok, here is the alternative to this PR: 

https://github.com/ceph/ceph/pull/35260/commits/9b4d9554121214623be4d5a5e6b7d9180acaf618

Both are ugly, but I have the impression that making the change here is a bit less intrusive.

**Comment by kshtsk** [comment]
> Both are ugly, but I have the impression that making the change here is a bit less intrusive.

It is mistaken feeling ;-) especially in the run methods :-D

**Comment by sebastian-philipp** [comment]
@kshtsk this PR is no longer a requirement. feel free to close this one now.

**Comment by kshtsk** [comment]
@susebot run deploy

**Pull request merged**